sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -0,0 +1,451 @@
+          case InfoLogLevel.FATAL_LEVEL | InfoLogLevel.ERROR_LEVEL => logError(_)
+          case InfoLogLevel.WARN_LEVEL | InfoLogLevel.INFO_LEVEL => logInfo(_)
+          case InfoLogLevel.DEBUG_LEVEL => logDebug(_)
+          case _ => logTrace(_)
+        }
+        loggingFunc(s"[Native-${infoLogLevel.getValue}] $logMsg")
+      }
+    }
+
+    var dbLogLevel = InfoLogLevel.ERROR_LEVEL
+    if (log.isWarnEnabled) dbLogLevel = InfoLogLevel.WARN_LEVEL
+    if (log.isInfoEnabled) dbLogLevel = InfoLogLevel.INFO_LEVEL
+    if (log.isDebugEnabled) dbLogLevel = InfoLogLevel.DEBUG_LEVEL
+    dbOptions.setLogger(dbLogger)
+    dbOptions.setInfoLogLevel(dbLogLevel)
+    logInfo(s"Set DB native logging level to $dbLogLevel")
+    dbLogger
+  }
+
+  /** Create a temp directory inside the local root directory */
+  private def createTempDir(prefix: String): File = {
+    Utils.createDirectory(localRootDir.getAbsolutePath, prefix)
+  }
+
+  /** Attempt to delete recursively, and log the error if any */
+  private def silentDeleteRecursively(file: File, msg: String): Unit = {
+    try {
+      Utils.deleteRecursively(file)
+    } catch {
+      case e: Exception =>
+        logWarning(s"Error recursively deleting local dir $file while $msg", e)
+    }
+  }
+
+  /** Records the duration of running `body` for the next query progress update. */
+  protected def timeTakenMs(body: => Unit): Long = Utils.timeTakenMs(body)._2
+
+  override protected def logName: String = s"${super.logName} $loggingId"
+}
+
+
+/** Mutable and reusable pair of byte arrays */
+class ByteArrayPair(var key: Array[Byte] = null, var value: Array[Byte] = null) {
+  def set(key: Array[Byte], value: Array[Byte]): ByteArrayPair = {
+    this.key = key
+    this.value = value
+    this
+  }
+}
+
+
+/**
+ * Configurations for optimizing RocksDB
+ *
+ * @param compactOnCommit Whether to compact RocksDB data before commit / checkpointing
+ */
+case class RocksDBConf(
+    minVersionsToRetain: Int,
+    compactOnCommit: Boolean,
+    pauseBackgroundWorkForCommit: Boolean,
+    blockSizeKB: Long,
+    blockCacheSizeMB: Long,
+    lockAcquireTimeoutMs: Long)
+
+object RocksDBConf {
+  /** Common prefix of all confs in SQLConf that affects RocksDB */
+  val ROCKSDB_CONF_NAME_PREFIX = "spark.sql.streaming.stateStore.rocksdb"
+
+  private case class ConfEntry(name: String, default: String) {
+    def fullName: String = s"$ROCKSDB_CONF_NAME_PREFIX.${name}".toLowerCase(Locale.ROOT)
+  }
+
+  // Configuration that specifies whether to compact the RocksDB data every time data is committed
+  private val COMPACT_ON_COMMIT_CONF = ConfEntry("compactOnCommit", "false")
+  private val PAUSE_BG_WORK_FOR_COMMIT_CONF = ConfEntry("pauseBackgroundWorkForCommit", "true")
+  private val BLOCK_SIZE_KB_CONF = ConfEntry("blockSizeKB", "4")
+  private val BLOCK_CACHE_SIZE_MB_CONF = ConfEntry("blockCacheSizeMB", "8")
+  private val LOCK_ACQUIRE_TIMEOUT_MS_CONF = ConfEntry("lockAcquireTimeoutMs", "60000")
```

Comment:
Note that we should document these configs at the end.

Reply:
Yea, I planned to document all the config for RocksDB in the user guide. If we have the configurable requirement in the future, people can easily be referenced.